### PR TITLE
WIP: Add dag for bqetl backfill

### PR DESCRIPTION
**Pull request by fbertsch**
I didn't have a chance to test this locally. I cobbled most of this together from @mikaeld's new backfill DAG, but clearly it's much simpler.

Hopefully I can get someone from DE to test this out, but I know that actually launching a job in the pod from local airflow is a real pain so it might not happen. LMK if SRE is opposed to yoloing this and iterating in prod.

### CODE REVIEW

**Comment by fbertsch** [comment]
@mikaeld any clue why this is failing on the param parsing? It's the same as the other backfill DAG.
```
[2022-12-06 17:06:16,196] {{dagbag.py:329}} ERROR - Failed to import: /app/dags/backfill-bqetl.py
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/airflow/models/dagbag.py", line 326, in _load_modules_from_file
    loader.exec_module(new_module)
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/app/dags/backfill-bqetl.py", line 104, in <module>
    dag = backfill_bqetl_dag()
  File "/usr/local/lib/python3.7/site-packages/airflow/models/dag.py", line 2332, in factory
    f(**f_kwargs)
  File "/app/dags/backfill-bqetl.py", line 87, in backfill_bqetl_dag
    params = __parse_string_params("{{ dag_run.conf }}")
  File "/app/dags/backfill-bqetl.py", line 37, in __parse_string_params
    params_parsed = ast.literal_eval(string_params)
  File "/usr/local/lib/python3.7/ast.py", line 91, in literal_eval
    return _convert(node_or_string)
  File "/usr/local/lib/python3.7/ast.py", line 76, in _convert
    return set(map(_convert, node.elts))
  File "/usr/local/lib/python3.7/ast.py", line 76, in _convert
    return set(map(_convert, node.elts))
  File "/usr/local/lib/python3.7/ast.py", line 90, in _convert
    return _convert_signed_num(node)
  File "/usr/local/lib/python3.7/ast.py", line 63, in _convert_signed_num
    return _convert_num(node)
  File "/usr/local/lib/python3.7/ast.py", line 55, in _convert_num
    raise ValueError('malformed node or string: ' + repr(node))
ValueError: malformed node or string: <_ast.Attribute object at 0x7f5b0f051410>
```

File: dags/backfill-bqetl.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    doc_md=doc_md,
+    catchup=False,
+    start_date=datetime.datetime(2022, 12, 1),
+    dagrun_timeout=datetime.timedelta(days=1),
+    tags=[Tag.ImpactTier.tier_3],
+    params={
+        "dataset_id": None,
+        "destination_table": None,
+        "dry_run": True,
+        "dry_run": True,
```
Comment (mikaeld) [inline dags/backfill-bqetl.py:71]: duplicate

File: dags/backfill-bqetl.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+def backfill_bqetl_dag():
+    param_validation_task = PythonOperator(
+        task_id = "param_validation",
+        python_callable = param_validation,
+        op_kwargs = {"params": "{{ dag_run.conf }}"},
+    )
+
+    params = __parse_string_params("{{ dag_run.conf }}")
+
+    backfill_task = bigquery_etl_backfill(
```
Comment (mikaeld) [inline dags/backfill-bqetl.py:89]: unresolved reference

File: dags/backfill-bqetl.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        "dataset_id": None,
+        "destination_table": None,
+        "dry_run": True,
+        "dry_run": True,
+        "end_date": datetime.date.today().isoformat(),
+        "excludes": [],
+        "no_partition": False,
+        "parallelism": 10,
+        "project_id": None,
+        "start_date": (datetime.date.today() - datetime.timedelta(days=10)).isoformat(),
```
Comment (mikaeld) [inline dags/backfill-bqetl.py:77]: use the new `Param` class (see current backfill.py implementation

**Review by mikaeld — CHANGES_REQUESTED** [review]

File: dags/backfill-bqetl.py (hunk 1 above)
Comment (fbertsch) [inline dags/backfill-bqetl.py:71]:
```suggestion
```

File: dags/backfill-bqetl.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    excludes: List[str]
+    dry_run: bool
+    parallelism: int
+    no_partition: bool
+
+    def validate_date_range(self) -> None:
+        start_date = datetime.datetime.fromisoformat(self.start_date)
+        end_date = datetime.datetime.fromisoformat(self.end_date)
+        if start_date > end_date:
+            raise ValueError(f"`start_date`={self.start_date} is greater than `end_date`={self.end_date}")
```
Comment (mikaeld) [inline dags/backfill-bqetl.py:27]: This should be in dags/utils/backfill.py

File: dags/backfill-bqetl.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    https://airflow.apache.org/docs/apache-airflow/stable/concepts/params.html
+    """
+    params_parsed = ast.literal_eval(string_params)
+    return BackfillParams(**params_parsed)
+
+
+def param_validation(params: str) -> bool:
+    backfill_params = __parse_string_params(params)
+    backfill_params.validate_date_range()
+    return True
```
Comment (mikaeld) [inline dags/backfill-bqetl.py:44]: This is no longer required; we are using Airflow 2.3 which supports the `Param` model, see current backfill DAG implementation.

File: dags/utils/gcp.py [hunk 6]
```diff
@@ -255,6 +256,88 @@ def bigquery_etl_query(
     )
 
 
+def bigquery_etl_backfill(
+        destination_table: str,
+        dataset_id: str,
+        start_date: str,
+        end_date: str = None,
+        project_id: str = None,
+        excludes: typing.List[str] = None,
```
Comment (mikaeld) [inline dags/utils/gcp.py:265]: types should be `Optional[str]` or use the new annotation syntax

```python
from __future__ import annotations
...
    end_date: str | None = None,
```

File: dags/utils/gcp.py [hunk 7]
```diff
@@ -255,6 +256,88 @@ def bigquery_etl_query(
...
+                                                   state, or the execution is interrupted.
+                                                   If False (default): do nothing, If True: delete the pod
+    :param Dict[str, Any] kwargs:                  Additional keyword arguments for
+                                                   GKEPodOperator
+    :return: GKEPodOperator
+    """
+    kwargs["task_id"] = kwargs.get("task_id", destination_table)
+    kwargs["name"] = kwargs.get("name", kwargs["task_id"].replace("_", "-"))
+    if not project_id:
+        project_id = "moz-fx-data-shared-prod"
```
Comment (mikaeld) [inline dags/utils/gcp.py:306]: this should be a default kwarg value instead of being tested here

File: dags/utils/gcp.py [hunk 8]
```diff
@@ -255,6 +256,88 @@ def bigquery_etl_query(
...
+        args += ["--end-date", end_date]
+    if excludes:
+        for exclude in excludes:
+            args += ["--exclude", exclude]
+    if dryrun:
+        args += ["--dry-run"]
+    if parallelism:
+        args += ["--parallelism", parallelism]
+    if no_partition:
+        args += ["--no-partition"]
```
Comment (mikaeld) [inline dags/utils/gcp.py:326]: `append` or `extend` methods are preferable

File: dags/utils/gcp.py [hunk 9]
```diff
@@ -255,6 +256,88 @@ def bigquery_etl_query(
...
+        f"{dataset_id}.{destination_table}",
+        "--start-date", start_date
+    ]
+
+    if end_date:
+        args += ["--end-date", end_date]
+    if excludes:
+        for exclude in excludes:
+            args += ["--exclude", exclude]
+    if dryrun:
```
Comment (mikaeld) [inline dags/utils/gcp.py:321]: unresolved reference `dryrun`, should be dry_run to match function signature.

**Review by mikaeld — CHANGES_REQUESTED** [review]

**Comment by mikaeld** [comment]
I've extracted the JSON pod definition from the error output. Error message states cannot unmarshal number into Go struct field Container.spec.containers.args of type string 

```json
{
  "apiVersion": "v1",
  "kind": "Pod",
  "metadata": {
    "annotations": {
      "cluster-autoscaler.kubernetes.io/safe-to-evict": "true"
    },
    "labels": {
      "dag_id": "backfill_bqetl",
      "task_id": "bigquery_etl_backfill",
      "run_id": "manual__2023-03-03T1915300000-23490cb28",
      "kubernetes_pod_operator": "True",
      "try_number": "1",
      "airflow_version": "2.3.3"
    },
    "name": "bigquery-etl-backfill-877d1ceecfa44462841182d1e6c9f173",
    "namespace": "default"
  },
  "spec": {
    "affinity": {},
    "containers": [
      {
        "args": [
          "script/bqetl",
          "query",
          "backfill",
          "google_ads_derived.campaign_conversions_by_date_v1",
          "--start-date",
          "2023-02-27",
          "--end-date",
          "2023-02-28",
          "--exclude",
          ",",
          "--parallelism",
          1
        ],
        "command": [],
        "env": [],
        "envFrom": [],
        "image": "gcr.io/moz-fx-data-airflow-prod-88e0/bigquery-etl:latest",
        "imagePullPolicy": "Always",
        "name": "base",
        "ports": [],
        "resources": {},
        "volumeMounts": []
      }
    ],
    "hostNetwork": false,
    "imagePullSecrets": [],
    "initContainers": [],
    "nodeSelector": {},
    "restartPolicy": "Never",
    "securityContext": {},
    "tolerations": [],
    "volumes": []
  }
}
```

the trailing `1` in `args` is definitely one of the issue. `args` is a string array but `1` of type `number`. Also, is the `","` arg necessary?

### Recommended fix
1. remove command generation code from bigquery_etl_backfill() , add it as a BackfillParams class method such as
```python
@dataclasses.dataclass
class BackfillParams:
    destination_table: str
    dataset_id: str
    start_date: str
    end_date: str
    project_id: str
    excludes: List[str]
    dry_run: bool
    parallelism: int
    no_partition: bool

    def validate_date_range(self) -> None:
        start_date = datetime.datetime.fromisoformat(self.start_date)
        end_date = datetime.datetime.fromisoformat(self.end_date)
        if start_date > end_date:
            raise ValueError(
                f"`start_date`={self.start_date} is greater than `end_date`={self.end_date}")

    def generate_backfill_command(self) -> list[str]:
        ...
```
2. Verify `generate_backfill_command` side effect using unit tests. Your command generation code is now decoupled, add unit tests. See `tests/utils/test_backfill.py` for example.
3. Use dependency injection. Use a `BackfillParams` instance as kwarg in `bigquery_etl_backfill`

**Comment by mikaeld** [comment]
@fbertsch any update

**Comment by fbertsch** [comment]
@mikaeld @ANich is going to pick this up

**Pull request closed without merging**